### PR TITLE
feat: 🎸 add end-to-end test harness

### PR DESCRIPTION
**Pull request by SSHari**
PR to add an end-to-end test harness using Playwright.

## Changes

1. Added a Github Action to run the end to end tests
2. Updated the `sendMagicLinkEmail` function to auto redirect to the magic link URL in test mode (Trying to parse the logs for the magic URL seemed unnecessary. We may consider just defaulting to this behavior in `development` in general.)
3. Added a test nextjs app with a basic job to use with the e2e tests (the values in `examples/nextjs-test/.env.example` are associated with the test data)
4. Added tests for creating an account and verifying jobs based on the test nextjs app from Change 3

## Running Locally

To run the tests locally, the following section from the publish github action can be followed:

```bash
# Setup environment variables
cp ./.env.example ./.env
cp ./examples/nextjs-test/.env.example ./examples/nextjs-test/.env.local
cp ./packages/database/.env.example ./packages/database/.env
pnpm run docker
pnpm run db:migrate
pnpm run db:seed
pnpm run test:e2e
pnpm run docker:stop
```

## Other

Since I can't verify the github action works until we merge this change into the main branch, I created a dummy github action with the same steps below that's triggered when my branch is updated. The results of that github action can be found here: https://github.com/SSHari/trigger.dev/actions/runs/5745732725/job/15574162405. One thing to note is that I couldn't run my test github action on the `buildjet-4vcpu-ubuntu-2204` runner, so I used the `ubuntu-22.04` instead.

Bounty: /claim #202

### CODE REVIEW

File: tests/e2e/e2e.spec.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { test, expect } from "@playwright/test";
+
+test("Create an account", async ({ page }) => {
+  await page.goto("http://localhost:3030/login");
```
Comment (zodman) [inline tests/e2e/e2e.spec.ts]: move the  URL to the `playwright.config.ts` use.baseURL. https://playwright.dev/docs/test-configuration

File: tests/e2e/e2e.spec.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  await page.getByLabel("Organization name").type("Test Org");
+  await page.getByLabel("Project name").type("Test Project");
+  await page.getByRole("button", { name: "Create" }).click();
+
+  await expect(page.locator("h1").filter({ hasText: /^Jobs$/ })).toBeVisible();
+});
+
+test("Verify jobs from the test nextjs project", async ({ page }) => {
+  await page.goto("http://localhost:3030");
```
Comment (zodman) [inline tests/e2e/e2e.spec.ts]: Same here

File: tests/e2e/e2e.spec.ts
Comment (zodman) [inline tests/e2e/e2e.spec.ts:1]: I suggest use the Page Object Pattern for getting test more easy to maintain...

https://www.selenium.dev/documentation/test_practices/encouraged/page_object_models/

File: playwright.config.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  forbidOnly: !!process.env.CI,
+  /* Retry on CI only */
+  retries: process.env.CI ? 2 : 0,
+  /* Opt out of parallel tests on CI. */
+  workers: process.env.CI ? 1 : undefined,
+  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
+  reporter: "html",
+  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
+  use: {
+    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
```
Comment (zodman) [inline playwright.config.ts:27]: here is missing the `baseURL` and readed from .env file

File: tests/utils.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  const prisma = new PrismaClient({
+    datasources: {
+      db: {
+        url: DATABASE_URL,
+      },
+    },
+  });
+
+  try {
```
Comment (zodman) [inline tests/utils.ts]: remove the try .. if the connection not success the test should be fail ... ASAP..

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -64,8 +64,54 @@ jobs:
       - name: 🔎 Type check
         run: pnpm run typecheck --filter webapp
 
+  e2e:
+    name: e2e Tests
+    runs-on: buildjet-4vcpu-ubuntu-2204
```
Comment (zodman) [inline .github/workflows/publish.yml:69]: here will be fail ? should be ubuntu-latest ?

**Review by zodman — COMMENTED** [review]
amazing!!!

I think is missing the document of  how to run the e2e in  `CONTRIBUITION.md` file 

give :money_with_wings:  to this PR!

File: apps/webapp/app/services/email.server.ts [hunk 6]
```diff
@@ -14,6 +15,11 @@ const client = new EmailClient({
 });
 
 export async function sendMagicLinkEmail(options: SendEmailOptions<AuthUser>): Promise<void> {
+  // Auto redirect when in test mode
+  if (env.APP_ENV === "test") {
```
Comment (zodman) [inline apps/webapp/app/services/email.server.ts:19]: i think this value is missing on the .env.example

File: apps/webapp/package.json [hunk 7]
```diff
@@ -14,6 +14,7 @@
     "start": "cross-env NODE_ENV=production node --max-old-space-size=8192 ./build/server.js",
     "typecheck": "tsc --noEmit",
     "db:seed": "node prisma/seed.js",
+    "db:seed:local": "ts-node prisma/seed.ts",
```
Comment (ericallam) [inline apps/webapp/package.json:17]: Good thinking 👍

File: apps/webapp/app/services/email.server.ts (hunk 6 above)
Comment (ericallam) [inline apps/webapp/app/services/email.server.ts]: Yea this makes a lot of sense. I think we should just do this when NODE_ENV=development

File: examples/nextjs-test/package.json [hunk 8]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "@examples/nextjs-test",
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "dev": "concurrently --kill-others \"pnpm dev:next\" \"wait-on http://localhost:3000 && pnpm dev:trigger\"",
+    "dev:next": "next dev",
+    "dev:trigger": "pnpm dlx @trigger.dev/cli@latest dev",
```
Comment (ericallam) [inline examples/nextjs-test/package.json]: This should use the locally built cli, just add `@trigger.dev/cli` to the dev dependencies like so:

```json
  "@trigger.dev/cli": "workspace:*"
```

And then update this script to `trigger-cli dev`

Will just need to add a `pnpm run build --filter @trigger.dev/cli` step to the e2e tests

**Review by ericallam — CHANGES_REQUESTED** [review]

**Comment by ericallam** [comment]
This is amazing, great work! Just a few tweaks and it should be ready to merge 👍

File: tests/e2e/e2e.spec.ts (hunk 1 above)
Comment (SSHari) [reply tests/e2e/e2e.spec.ts]: I've added the base URL to the config file and updated the tests to use the relative paths.

File: tests/e2e/e2e.spec.ts (hunk 2 above)
Comment (SSHari) [reply tests/e2e/e2e.spec.ts]: I've added the base URL to the config file and updated the tests to use the relative paths.

File: playwright.config.ts (hunk 3 above)
Comment (SSHari) [reply playwright.config.ts:27]: I've added the base URL to the config file and updated the tests to use the relative paths.

File: tests/utils.ts (hunk 4 above)
Comment (SSHari) [reply tests/utils.ts]: Removed the `try/catch`

File: apps/webapp/app/services/email.server.ts (hunk 6 above)
Comment (SSHari) [reply apps/webapp/app/services/email.server.ts]: I've updated this to auto-redirect based on `NODE_ENV === "development"`

File: examples/nextjs-test/package.json (hunk 8 above)
Comment (SSHari) [reply examples/nextjs-test/package.json]: I've updated this script to run `trigger-cli dev` directly.

**Review by matt-aitken — APPROVED** [review]
This is awesome @SSHari, merging now. Thanks!

**Pull request merged**